### DISCUSSION TITLE
pixel color palette control

### DISCUSSION THREAD

**Discussion by joseluis** [dankamongmen/notcurses#1512] [Ideas]
Continuing the conversation from https://github.com/dankamongmen/notcurses/discussions/1488#discussioncomment-582398 in order to focus on sixel/pixel color control...

> btw, details of how sixel works mean that you're not necessarily going to get this level of control anyway without more work. are you assuming that if, say, there are 256 color registers, you can perfectly reproduce 256 different 8bpc RGB values? no sir; sixel works on a [0..100] range for RGB. so if you had FFFFFF and FEFEFE, guess what, those are going to both be mapped to 100,100,100 (and thus be, and count as, a single color).

Whatever we can know, we should be able to know... I'd like to understand better the mechanics of the reduction of the color range. I guess it's not upper bounded, so it would happen too with 222222 and 232323, right? As I understand each 0-255 value gets transposed to 0-100?

It would be useful to have a function to convert an rgb color to what that color would be in a sixel representation...

**Comment by dankamongmen** [comment]
https://github.com/dankamongmen/notcurses/issues/1391 is useful background information, with some input from the fabulous @dnkl

**Reply by dnkl** [reply]
I don't think I've _ever_ been called "fabulous" before :D

Not sure I have anything to contribute with here though. Except perhaps just mentioning (again, sorry if you're already well aware of all this) that you're not limited to 256 colors; using `\E[?1;3;1024S` (for example), you can bump it to 1024 colors.

Use `\E[?1;1;0S` to get the **current** palette size, and `\E[?1;4;0S` to get the **maximum** palette size.

**Reply by dankamongmen** [reply]
aye, we probably ought definitely take it up to the maximum reported. this will admittedly allow sixel color quantization to take longer (if the total number of colors is below the lower old value, the time ought not perceptibly change).

**Comment by dankamongmen** [comment]
> Whatever we can know, we should be able to know... I'd like to understand better the mechanics of the reduction of the color range. I guess it's not upper bounded, so it would happen too with 222222 and 232323, right? As I understand each 0-255 value gets transposed to 0-100?

the color quantization is highly dependent on implementation. you can encode to *either* 0..100 RGB *or* 0..100 HSL. currently, notcurses uses rgb exclusively, and first does a flat linear conversion of RGB 24bpp colors to RGB umm 19.92bpp, fuck sixel right in the ass. but i bet there are algorithms which do something different, making better use of the original precision.

> It would be useful to have a function to convert an rgb color to what that color would be in a sixel representation...

no such function exists, except in the case of the very simplest color quantizers, because the colors are sensitive to the entire image, not just individual pixels.
